title: "DADA2_16S"
author: "Johanna"
date: "2025-02-27"
output: html_document
---

```{bash}
conda activate 18S_
```

# DADA2 Pipeline for 16S V3V4 rDNA sequencing reads
In this script, I perform the DADA2 pipeline apply Malte's DADA2 Workflow (https://github.com/mruehlemann/ikmb_amplicon_processing/blob/master/dada2_16S_workflow_with_AR.R) on my 18S V4V5 rDNA sequencing reads. 
All steps are performed in R. 

```{bash}
R
```
# load libraries

```{r}
library(tidyverse) 
library(phyloseq) 
library(microbiome)
library(DESeq2)
library(dada2)
library(vegan)
library(readxl)
 
#set your paths
path_reads <- "/work_beegfs/sukmb626/rawdata/16S_Raw_Reads"       #where the raw reads are stored
```
## Raw Reads 
First, raw sequencing reads (in FastQ Files) are loaded in the R environment, 
forward and reverse reads are kept, and quality profiles are plotted. 

```{r}
fns <- list.files(path_reads)            # list all files

fastqs <- fns[grepl(".fastq.gz$", fns)]  # keep only FastQ files
fastqs <- sort(fastqs)                   # Sort ensures forward/reverse reads are in same order

# make sure that R1 is for forward read and R2 for reverse
fnFs <- fastqs[grepl("R1_001.fastq.gz", fastqs)] ## Just the forward read files
fnRs <- fastqs[grepl("R2_001.fastq.gz", fastqs)] ## Just the reverse read files

sample.names <- sapply(strsplit(fnFs, "_"), `[`, 2)  ## Sample names are derived from the filenames  
cat(paste0("Starting processing for ",length(sample.names)," samples\n"))
duplicated(sample.names)

fnFs_full <- file.path(path_reads, fnFs)  ## Fully specify the path for the fnFs 
fnRs_full <- file.path(path_reads, fnRs)  ## Fully specify the path for the fnRs

# Quality profiles & save the Rplots

QualityPlot <- plotQualityProfile(c(fnFs_full[1:2],fnRs_full[1:2]))   # Quality profiles
pdf("/home/sukmb626/Madagscar/16S/QualityPlot.pdf")                   # save the Rplots in the home directory 
dev.off()

setwd(path_wd)

outdir <- paste0("data")
```

## Trimming and Filterung of Forward and Reverse Reads
Intermediated, filtered files will be stored in an extra folder, where the Trimming and Filtering is performed. 

```{r}
###check if there are any duplicates
duplicated(c(filtFs, filtRs))   
out <- filterAndTrim(fnFs_full, filtFs, fnRs_full, filtRs, truncLen=c(250,210),trimLeft=c(5, 5), maxN=0, maxEE=c(2,2), truncQ=5, rm.phix=TRUE, compress=TRUE, multithread=4) #save the filtered and timmed reads
### check the dimentions of out1
dim (out)  

### check reads after trimming
plotQualityProfile(c(filtFs[1:2], fnFs_full[1:2])) #Check Forward reads after trimming 
plotQualityProfile(c(filtRs[1:2], fnRs_full[1:2])) #Check Reverse reads after trimming 
```

## Learning Error Rates
In an output directory, error models will be stored. 100 million bases are used to estimate error rates and the error rates will be plotted for forward and reverse reads. 
```{r}
### create output directory
dir.create(paste0(outdir,"/errors"),recursive=T,showWarnings=F)  #create output directory


### Learn forward error rates
errF <- learnErrors(filtFs, nbases=100000000, multithread=4)
saveRDS(errF, paste0(outdir,"/errors/errF.Rds"))

### Learn reverse error rates
errR <- learnErrors(filtRs, nbases=100000000, multithread=4)
saveRDS(errR, paste0(outdir,"/errors/errR.Rds"))

plotErrors(errF, nominalQ=TRUE)
plotErrors(errR, nominalQ=TRUE)

###save the plot
png("myplot.png")
dev.off()
```

## Dereplication 
Collapsing identical sequences to speed up the processing
```{r}
filtFs <- filtFs[file.exists(filtFs)]
filtRs <- filtRs[file.exists(filtRs)]

### remove any files, that don't exist
derepFs <- derepFastq(filtFs, verbose=FALSE)
derepRs <- derepFastq(filtRs, verbose=FALSE)
```

## Sequence Inference
Denoising and Merging of reads is performed. 
```{r}
###Denoising: Identifies real biological sequences while removing sequencing errors.
dadaFs <- dada(derepFs, err=errF, multithread=4)
dadaRs <- dada(derepRs, err=errR, multithread=4)

### Merging of forward and reverse reads into full sequences 
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs)
```

## Chimera Removal
Tables of unique sequencies and the removal of chimeric sequencies is performed here and 
the seqencing tables are saved as RDS files. 
```{r}
seqtab <- makeSequenceTable(mergers)
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=4, verbose=TRUE)

### safe the seqtab.nochim
saveRDS(seqtab.nochim, "/home/sukmb626/seqtab.nochim_16S_all.rds")
saveRDS(seqtab, "/home/sukmb626/seqtab16S_all.rds")
```
